nixd/lspserver/src/Connection.cpp

```diff
@@ -222,24 +223,53 @@ bool InboundPort::readStandardMessage(std::string &JSONString) {
 bool InboundPort::readDelimitedMessage(std::string &JSONString) {
   JSONString.clear();
   llvm::SmallString<128> Line;
-  bool IsInputBlock = false;
+  bool IsJSONBlock = false;
+  std::optional<std::string> NixDocURI;
+  std::string NixDoc;
   while (readLine(In, Close, Line)) {
     auto LineRef = Line.str().trim();
-    if (IsInputBlock) {
-      // We are in input blocks, read lines and append JSONString.
+    if (IsJSONBlock) {
+      // We are in a JSON block, read lines and append JSONString.
       if (LineRef.starts_with("#")) // comment
         continue;
 
       // End of the block
       if (LineRef.starts_with("```")) {
-        IsInputBlock = false;
+        IsJSONBlock = false;
         break;
       }
 
       JSONString += Line;
+    } else if (NixDocURI.has_value()) {
+      // We are in a Nix block. (This was implemented to make the .md test
+      // files more readable, particularly regarding multiline Nix documents,
+      // so that the newlines don't have to be \n escaped.)
+
+      if (LineRef.starts_with("```")) {
+        nlohmann::json DidOpen = {
+            {"jsonrpc", "2.0"},
+            {"method", "textDocument/didOpen"},
+            {"params",
+             {{"textDocument",
+               {
+                   {"uri", NixDocURI.value()},
+                   {"languageId", "nix"},
+                   {"version", 1},
+                   {"text", llvm::StringRef(NixDoc).rtrim()},
+               }}}},
+        };
+        JSONString += DidOpen.dump();
```

Comment:
Here, it seems that we are generating a JSON file in code and then using another piece of code to parse it, which feels somewhat inefficient. I did take a look locally, and it seems that modifying [nixd/lspserver/src/Connection.cpp#L293](https://github.com/nix-community/nixd/blob/3aa27fde1edcf7b126c70a62aad05d120209363c/nixd/lspserver/src/Connection.cpp#L266) would be a bit tricky. Still, I'd prefer not to have our structured data serialized into a string only to be parsed again by our own program. Do you have any thoughts on this?

Reply:
This is just test code so I wouldn't worry about the performance. But sure, refactored it to move the JSON parsing into the format-specific read functions. Note that to still log the incoming messages in the main loop they are now serialized again.